src/loader/loader.c

```diff
@@ -670,6 +669,23 @@ static void inline do_load()
 	char soname[MAX_SO_NAME_LEN];
 	post_parse_analyze_hook_type old_hook;
 
+	/* If the right version of the library is already loaded, we will just
+	 * skip the actual loading. If the wrong version of the library is loaded,
+	 * we need to kill the session since it will not be able to continue
+	 * operate. */
+	if (*soversion)
+	{
+		if (strcmp(soversion, version) == 0)
+			return;
+		ereport(FATAL,
+				(errcode(ERRCODE_DUPLICATE_OBJECT),
```

Comment:
Can we add some test here???

Reply:
No, that is unfortunately not possible since it requires upgrading between different versions of the server. The problem only manifests itself when upgrading from one version of the server to a different version of the server.